stacks-common/src/util/vrf.rs

```diff
@@ -149,71 +114,76 @@ impl Default for VRFPrivateKey {
-        VRFPrivateKey(keypair.secret)
+        let signing_key = ed25519_dalek::SigningKey::generate(&mut rng);
+        VRFPrivateKey(signing_key)
     }
 
     pub fn from_hex(h: &str) -> Option<VRFPrivateKey> {
-        match hex_bytes(h) {
-            Ok(b) => match ed25519_PrivateKey::from_bytes(&b[..]) {
-                Ok(pk) => Some(VRFPrivateKey(pk)),
-                Err(_) => None,
-            },
-            Err(_) => None,
-        }
+        let bytes = hex_bytes(h).ok()?;
+        Self::from_bytes(bytes.as_slice())
     }
 
     pub fn from_bytes(b: &[u8]) -> Option<VRFPrivateKey> {
-        match ed25519_PrivateKey::from_bytes(b) {
-            Ok(pk) => Some(VRFPrivateKey(pk)),
-            Err(_) => None,
-        }
+        let signing_key = ed25519_dalek::SigningKey::try_from(b).ok()?;
+        Some(VRFPrivateKey(signing_key))
     }
 
     pub fn to_hex(&self) -> String {
         to_hex(self.as_bytes())
     }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+
+    pub fn to_bytes(&self) -> [u8; 32] {
+        self.0.to_bytes()
+    }
 }
 
 impl VRFPublicKey {
-    pub fn from_private(pk: &VRFPrivateKey) -> VRFPublicKey {
-        VRFPublicKey(ed25519_PublicKey::from(&pk.0))
+    pub fn from_private(sk: &VRFPrivateKey) -> VRFPublicKey {
+        VRFPublicKey(sk.0.verifying_key())
     }
 
+    /// Verify that a given byte string is a well-formed EdDSA public
+    /// key (i.e. it's a compressed Edwards point that is valid), and return
+    /// a VRFPublicKey if so
     pub fn from_bytes(pubkey_bytes: &[u8]) -> Option<VRFPublicKey> {
-        match pubkey_bytes.len() {
-            32 => {
-                let mut pubkey_slice = [0; 32];
-                pubkey_slice.copy_from_slice(&pubkey_bytes[0..32]);
-
-                let checked_pubkey = CompressedEdwardsY(pubkey_slice);
-                match checked_pubkey.decompress() {
-                    Some(_) => {}
-                    None => {
-                        // invalid
-                        return None;
-                    }
-                }
-
-                match ed25519_PublicKey::from_bytes(&pubkey_slice) {
-                    Ok(key) => Some(VRFPublicKey(key)),
-                    Err(_) => None,
-                }
-            }
-            _ => None,
+        let pubkey_slice = pubkey_bytes.try_into().ok()?;
+
+        // NOTE: `ed25519_dalek::VerifyingKey::from_bytes` docs say
+        //  that this check must be performed by the caller, but as of
+        //  latest, it actually performs the check as well. However,
+        //  we do this check out of an abundance of caution because
+        //  that's what the docs say to do!
+
+        let checked_pubkey = CompressedEdwardsY(pubkey_slice);
+        if checked_pubkey.decompress().is_none() {
+            // invalid
+            return None;
         }
+
+        let key = ed25519_dalek::VerifyingKey::from_bytes(&pubkey_slice).ok()?;
+        Some(VRFPublicKey(key))
```

Comment:
Simpler way to write this:
```suggestion
        ed25519_dalek::VerifyingKey::from_bytes(&pubkey_slice).map(VRFPublicKey).ok()
```

Reply:
I'm a little conflicted about this kind of thing -- it's actually not obvious to me which of the above is simpler. Sometimes the functional chaining in `option` and `result` can be hard to decipher (especially because it often means that the `Err` type survives to the end of the chain).